### PR TITLE
WT-10273: First draft of new RTS sections

### PR DESCRIPTION
**Pull request by korteland**
This is a draft version right now. The main things I want to check for are (1) the accuracy of the content, and (2) that it's covering useful topics.

I'll format it for Doxygen and move it into the existing page once we're done reviewing for technical content since I find Markdown easier to read and write.

### CODE REVIEW

File: src/docs/rts-new.md [hunk 1]
```diff
@@ -0,0 +1,170 @@
+# Interaction with timestamps
+
+The main thing RTS is concerned with are the durable and stable
+timestamps. The high-level details are covered above, what follows is a more
+detailed view of how we use these timestamps.
+
+We read `txn_global` for pinned and stable timestamps at the start of RTS,
```
Comment (sueloverso) [inline src/docs/rts-new.md:7]: Is `pinned` discussed above? Because just reading this I expected you to talk about durable and stable timestamps from the previous paragraph. Now durable is not mentioned but pinned is introduced.

I suggest not referring to `txn_global` by structure variable name. It is not mentioned anywhere else in the docs. I suggest rewording to `We read the global pinned and stable timestamps...`

File: src/docs/rts-new.md [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+and in diagnostic builds we read them again at the end to assert they didn't
+change underneath us (indicating the system was not quiescent, and therefore
+a bug).
+
+Once RTS has finished, we roll back the global durable timestamp to the
+stable timestamp, since if there are no unstable updates, `txn_global`'s
+durable timestamp can by definition be moved backwards to the stable
+timestamp. If we weren't to do this, it wouldn't automatically be a problem,
+but we'd likely end up unnecessarily scanning tables the next time RTS is
+run.
```
Comment (sueloverso) [inline src/docs/rts-new.md:17]: The last sentence does not add much to me. "it wouldn't automatically be a problem" sounds like it may sometimes be a problem. I suggest just removing it and stating what we do.

File: src/docs/rts-new.md [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+detailed view of how we use these timestamps.
+
+We read `txn_global` for pinned and stable timestamps at the start of RTS,
+and in diagnostic builds we read them again at the end to assert they didn't
+change underneath us (indicating the system was not quiescent, and therefore
+a bug).
+
+Once RTS has finished, we roll back the global durable timestamp to the
+stable timestamp, since if there are no unstable updates, `txn_global`'s
+durable timestamp can by definition be moved backwards to the stable
```
Comment (sueloverso) [inline src/docs/rts-new.md:14]: Minor: "...durable timestamp can, by definition, be..." (add the commas)

File: src/docs/rts-new.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+  timestamp. Thus, we can have an update in the history store with a maximum
+  stop timestamp in the middle of other updates for the same key.
+  
+We also have some verification specifically related to the history
+store. While iterating over records, we check that the start time of a newly
+encountered record is greater than or equal to the previous record. We also
+validate that the timestamps in the key and the cell are the same (where
+possible).
+
+Fast-truncate is worth mentioning briefly. If the page has a `page_del`
```
Comment (sueloverso) [inline src/docs/rts-new.md:42]: Again, don't use the name of the code structure. Just say "If the page has a page delete structure..."

File: src/docs/rts-new.md [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+structure, we can look at its `durable_timestamp`, much the same as we would
```
Comment (sueloverso) [inline src/docs/rts-new.md:43]: I'd separate durable timestamp into just words.

File: src/docs/rts-new.md [hunk 6]
```diff
@@ -0,0 +1,170 @@
+# Interaction with timestamps
```
[the remark this answers is not in the document]
Comment (korteland) [reply src/docs/rts-new.md:1]: Yep, that's right. I find Markdown easier to deal with so it's living in here while I still expect a lot of changes.

File: src/docs/rts-new.md [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+
+The main thing RTS is concerned with are the durable and stable
+timestamps. The high-level details are covered above, what follows is a more
+detailed view of how we use these timestamps.
```
[the remark this answers is not in the document]
Comment (korteland) [reply src/docs/rts-new.md:5]: Yes, it would get added to the bottom of `arch-rts.dox`. Sorry that wasn't clear from the description.

File: src/docs/rts-new.md [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+
+This is a little simpler than the timestamp rules - we use normal snapshot
+visibility in most of the places we look at timestamps (see the timestamps
+section). One exception is that we treat all data as visible when RTS is
+called from the API (as opposed to startup or shutdown). 
+
+We also clear the transaction ID of updates we look at during RTS, because
+the connection's write generation will be initialised after RTS and the
+updates in the cache would be problematic for other parts of the code if
+they had a "legitimate" transaction ID.
```
Comment (korteland) [inline src/docs/rts-new.md:56]: @kommiharibabu it wasn't clear from the code why this is the case, do you know why?

File: src/docs/rts-new.md [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+
+Fast-truncate is worth mentioning briefly. If the page has a `page_del`
+structure, we can look at its `durable_timestamp`, much the same as we would
+a normal page.
+
+# Interaction with transaction IDs
+
+This is a little simpler than the timestamp rules - we use normal snapshot
+visibility in most of the places we look at timestamps (see the timestamps
+section). One exception is that we treat all data as visible when RTS is
```
Comment (korteland) [inline src/docs/rts-new.md:50]: @kommiharibabu the reason behind this also wasn't clear to me from the code, do you know anything about why we treat it like this?

File: src/docs/rts-new.md [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+RTS tries to avoid looking at a page if it can. There are two ways we can
+avoid instantiating a page: if it was truncated at/before the stable
+timestamp, and if the page state is anything other than on-disk. If neither
+of these are satisfied, we need to instantiate the page to look at it, but
+that doesn't mean we look at the contents. Rather, we use the existing tree
+walk custom skipping functionality to check the page metadata, stuff like
+reconciliation results. These checks are fairly low-level, so they're
+outside the scope of this document.
+
+One final note is that we deliberately do not call RTS for internal pages.
```
Comment (korteland) [inline src/docs/rts-new.md:105]: @kommiharibabu I suspect this is because they won't have "updates" to roll back as such, and we handle tree maintenance (node merging, splitting etc) during reconciliation. Is this correct? If so, I'll add a short explanation.

File: src/docs/rts-new.md [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+
+Once RTS has finished, we roll back the global durable timestamp to the
+stable timestamp, since if there are no unstable updates, `txn_global`'s
+durable timestamp can by definition be moved backwards to the stable
+timestamp. If we weren't to do this, it wouldn't automatically be a problem,
+but we'd likely end up unnecessarily scanning tables the next time RTS is
+run.
+
+Timestamps of updates in the history store can have some surprising
+properties:
```
[the remark this answers is not in the document]
Comment (korteland) [reply src/docs/rts-new.md:18]: No, there's no other document here, except maybe `arch-rts.dox` (which this will get appended to). The definition of "surprising" was "stuff Will was surprised by", so I've changed it to "non-obvious".

(My goal with this section is to explain some of the weirder scenarios it's possible to run into - everyone will have a different definition of weird, but I figured something is better than nothing.)

File: src/docs/rts-new.md [hunk 12]
```diff
@@ -0,0 +1,170 @@
+# Interaction with timestamps
+
+The main thing RTS is concerned with are the durable and stable
```
[the remark this answers is not in the document]
Comment (korteland) [reply src/docs/rts-new.md]: Thanks, fixed in 4f053fe.

File: src/docs/rts-new.md (hunk 1 above)
Comment (korteland) [reply src/docs/rts-new.md]: Thanks, fixed in 4f053fe.

File: src/docs/rts-new.md (hunk 2 above)
Comment (korteland) [reply src/docs/rts-new.md]: Thanks - I've reworded this section so that the motivation behind rolling back the global durable timestamp is clearer in 4f053fe.

File: src/docs/rts-new.md (hunk 3 above)
Comment (korteland) [reply src/docs/rts-new.md]: Thanks, fixed in 4f053fe.

File: src/docs/rts-new.md [hunk 13]
```diff
@@ -0,0 +1,170 @@
...
+but we'd likely end up unnecessarily scanning tables the next time RTS is
+run.
+
+Timestamps of updates in the history store can have some surprising
+properties:
+
+* The start timestamp can be equal to the stop timestamp if the original
+  update's commit timestamp is in order.
+* During rollback of a prepared transaction, a history store update with a
+  stop timestamp may not get removed, leading to a duplocate record. RTS
```
[the remark this answers is not in the document]
Comment (korteland) [reply src/docs/rts-new.md]: Thanks, fixed in 4f053fe.

File: src/docs/rts-new.md (hunk 4 above)
Comment (korteland) [reply src/docs/rts-new.md]: Thanks, fixed in 4f053fe.

File: src/docs/rts-new.md (hunk 5 above)
Comment (korteland) [reply src/docs/rts-new.md]: Thanks, fixed in 4f053fe.

File: src/docs/rts-new.md (hunk 9 above)
Comment (kommiharibabu) [reply src/docs/rts-new.md:48]: The need to maintain multiple version updates in the database is to support an older reader that cannot see the newer modifications. Whenever there are no older readers we do not need to maintain any older versions as the latest version is visible to all the newer transactions.

The condition to call WiredTiger rollback to stable is to stop all the running transactions in the database. So as per the transaction rules, the latest update is stable. Due to this reason, we do not verify the transaction stability as it is not required.

File: src/docs/rts-new.md (hunk 10 above)
Comment (kommiharibabu) [reply src/docs/rts-new.md:103]: Yes, we do not touch the internal pages as they do not have updates. Just to be clear, we read the internal pages into the cache to abort any unstable updates on the leaf pages of the internal pages.

File: src/docs/rts-new.md [hunk 14]
```diff
@@ -0,0 +1,170 @@
...
+We read `txn_global` for pinned and stable timestamps at the start of RTS,
+and in diagnostic builds we read them again at the end to assert they didn't
+change underneath us (indicating the system was not quiescent, and therefore
+a bug).
+
+Once RTS has finished, we roll back the global durable timestamp to the
+stable timestamp, since if there are no unstable updates, `txn_global`'s
+durable timestamp can by definition be moved backwards to the stable
+timestamp. If we weren't to do this, it wouldn't automatically be a problem,
+but we'd likely end up unnecessarily scanning tables the next time RTS is
```
[the remark this answers is not in the document]
Comment (korteland) [reply src/docs/rts-new.md]: Thanks - I went looking for code to back this claim up and couldn't find any, so I probably just got confused when I originally wrote this section. Fixed in ae692da.

File: src/docs/rts-new.md (hunk 8 above)
Comment (korteland) [reply src/docs/rts-new.md:51]: Thanks - I've added the example we discussed in ae692da.

File: src/docs/rts-new.md (hunk 10 above)
Comment (korteland) [reply src/docs/rts-new.md]: Thanks for confirming, I've added some details to this effect in ae692da.

File: src/docs/rts-new.md [hunk 15]
```diff
@@ -0,0 +1,178 @@
...
+normal page.
+
+# Interaction with transaction IDs
+
+This is a little simpler than the timestamp rules - we use normal snapshot
+visibility in most of the places we look at timestamps (see the timestamps
+section).
+
+We also clear the transaction ID of updates we look at during RTS, because
+the connection's write generation will be initialised after RTS and the
```
Comment (sueloverso) [inline src/docs/rts-new.md:49]: Use American spelling: `initialized`

File: src/docs/rts-new.md [hunk 16]
```diff
@@ -0,0 +1,178 @@
...
+are quite involved, and out of scope for the architecture guide.
+
+# Interaction with checkpoints and eviction
+
+Checkpoints cannot be performed while RTS is running, even if they start
+before RTS. This is part of the contract that the system must be in a
+quiescent state before running RTS, but it's also enforced by taking the
+session's checkpoint lock.
+
+Upon completion of RTS, a checkpoint is performed to that both in-memory and
```
Comment (sueloverso) [inline src/docs/rts-new.md:134]: Minor "to" -> "so"

File: src/docs/rts-new.md [hunk 17]
```diff
@@ -0,0 +1,178 @@
...
+on-disk versions of data are the same. It's possible for users to opt out of
+this, for example if they intended to do a checkpoint shortly after RTS
+anyway.
+
+Eviction doesn't directly interact with RTS in any special way, but it's
+worth pointing out that RTS is likely to generate a large number of both
+clean and dirty pages, causing some amount of extra cache pressure. The
+clean pages aren't a big deal, but the dirty pages will need to be written
```
Comment (sueloverso) [inline src/docs/rts-new.md:142]: "aren't a big deal" should be something more concrete and technical like "can be discarded from cache" or something.

File: src/docs/rts-new.md [hunk 18]
```diff
@@ -0,0 +1,178 @@
...
+out, causing a potentially large amount of I/O and the potential for
+application threads to be forced into doing eviction work. Eviction triggers
+and targets apply the same way during RTS as they do at any other time.
```
Comment (sueloverso) [inline src/docs/rts-new.md:145]: So the end result is that RTS causes cache pressure and that's just the way it is. There is nothing for the user or system to do here.

File: src/docs/rts-new.md [hunk 19]
```diff
@@ -0,0 +1,178 @@
...
+
+# Dry-run mode
+
+Dry-run mode is a way of running RTS without making any changes to on-disk
+or in-memory data structures, including the history store. It uses the same
+code as "normal" RTS wherever possible, making it ideal for checking what
```
Comment (sueloverso) [inline src/docs/rts-new.md:151]: Avoid quoted words like "normal" above. I suggest just removing the quotes

File: src/docs/rts-new.md [hunk 20]
```diff
@@ -0,0 +1,178 @@
...
+RTS would do on a given system, without disturbing it. One limitation is
+that it stills needs to take the same locks as real RTS, so the system must
+still be quiescent.
+
+Dry-run mode does not, however, increment all of the same statistics as
+"normal" RTS. As a rule, it only updates statistics that reflect what it's
```
Comment (sueloverso) [inline src/docs/rts-new.md:157]: Here too

File: src/docs/rts-new.md [hunk 21]
```diff
@@ -0,0 +1,178 @@
...
+doing. For example, it will increment statistics about the number of trees
+and pages visited, but it will not change any statistics about the number of
+updates aborted (since it is doing real work visiting pages, but it isn't
+aborting any unstable updates).
+
+Dry-run mode will cause similar cache pressure to the real thing, since it
```
Comment (sueloverso) [inline src/docs/rts-new.md:163]: "the real thing" is too casual. I suggest `normal mode`.

File: src/docs/rts-new.md [hunk 22]
```diff
@@ -0,0 +1,178 @@
...
+must visit the same trees and pages. There is a slight difference around
+eviction though, since it will not mark ny pages as dirty, meaning eviction
```
Comment (sueloverso) [inline src/docs/rts-new.md:165]: Typo: `ny` -> `any`

File: src/docs/rts-new.md [hunk 23]
```diff
@@ -0,0 +1,178 @@
...
+should have a much easier time cleaning up after a dry run.
+
+# `rollback_to_stable_one`
```
Comment (sueloverso) [inline src/docs/rts-new.md:168]: We should avoid function names as headers. How about "RTS for a single table"

File: src/docs/rts-new.md [hunk 24]
```diff
@@ -0,0 +1,178 @@
...
+
+There is an internal form of rollback to stable which can operate on a
+single object, called `rollback_to_stable_one`. The API requires a URI is
+passed in, which contrasts to `rollback_to_stable` where it will
+deliberately look at every single tree in the database.
+
+This is primarily used for salvage. Salvage accepts a single URI to attempt
+fixing, so we don't want "regular" RTS to potentially read everything,
```
Comment (sueloverso) [inline src/docs/rts-new.md:176]: Again removed quotes

**Review by sueloverso — APPROVED** [review]
I have a bunch of minor text suggestions. But you don't need to wait for me to review again unless you change a bunch more. This is a great addition. I cannot say I understand all of it right away but it is helpful.

File: src/docs/rts-new.md [hunk 25]
```diff
@@ -0,0 +1,178 @@
...
+timestamp can, by definition, be moved backwards to the stable timestamp.
+
+Timestamps of updates in the history store can have some non-obvious
+properties:
+
+* The start timestamp can be equal to the stop timestamp if the original
+  update's commit timestamp is in order.
+* During rollback of a prepared transaction, a history store update with a
+  stop timestamp may not get removed, leading to a duplicate record. RTS
+  ignores these.
```
[the remark this answers is not in the document]
Comment (korteland) [reply src/docs/rts-new.md]: Good idea, done in 528be5c.

File: src/docs/rts-new.md (hunk 15 above)
Comment (korteland) [reply src/docs/rts-new.md]: Normally I say thanks or something, but for this one I'm just going to passive-aggressively link to the fix ;-) 5ec1700

File: src/docs/rts-new.md (hunk 16 above)
Comment (korteland) [reply src/docs/rts-new.md]: Good spot, fixed in 5ec1700.

File: src/docs/rts-new.md (hunk 18 above)
Comment (korteland) [reply src/docs/rts-new.md:145]: Yep, just adding some explanation for why it happens.

File: src/docs/rts-new.md (hunk 19 above)
Comment (korteland) [reply src/docs/rts-new.md]: Yep, thanks - fixed in 5ec1700.

File: src/docs/rts-new.md (hunk 20 above)
Comment (korteland) [reply src/docs/rts-new.md]: Thanks, fixed in 5ec1700.

File: src/docs/rts-new.md (hunk 21 above)
Comment (korteland) [reply src/docs/rts-new.md]: Yep, thanks - fixed in 5ec1700.

File: src/docs/rts-new.md (hunk 23 above)
Comment (korteland) [reply src/docs/rts-new.md]: Thanks - I prefer this as well. Fixed in 5ec1700.

File: src/docs/rts-new.md (hunk 24 above)
Comment (korteland) [reply src/docs/rts-new.md]: Thanks, fixed in 5ec1700.

File: src/docs/rts-new.md (hunk 15 above)
Comment (sueloverso) [reply src/docs/rts-new.md]: Well, all I can say is that when Michael and Keith started WT they made a conscious decision to standardize on that.

File: src/docs/arch-rts.dox [hunk 26]
```diff
@@ -110,4 +109,154 @@ Update list: U3 (30)\n
...
+Less obviously, the history store needs one final inspection regardless of whether any other B-Trees
+were touched. This is because we usually rely on getting to the history store "via" the data store,
+but if we don't touch a B-Tree, that doesn't mean there aren't updates in the history store after
+the stable timestamp.
+
+@section iterating-over-pages Iterating over pages
+
+RTS tries to avoid looking at a page if it can. There are two ways we can avoid instantiating a
+page: if it was truncated at/before the stable timestamp, and if the page state is anything other
+than on-disk. If neither of these are satisfied, we need to instantiate the page to look at it, but
```
Comment (kommiharibabu) [inline src/docs/arch-rts.dox:198]: If the page state is other than on-disk means it is already in the cache, so there is no instantiation here. Are you explaining the page with the on-disk state that does not have unstable updates will not be instantiated?

File: src/docs/arch-rts.dox [hunk 27]
```diff
@@ -110,4 +109,154 @@ Update list: U3 (30)\n
...
+reconcile it in future.
+
+The mechanics of deleting items (or inserting tombstones) from these places are quite involved, and
+out of scope for the architecture guide.
+
+@section interaction-with-checkpoints-and-eviction Interaction with checkpoints and eviction
+
+Checkpoints cannot be performed while RTS is running, even if they start before RTS. This is part of
+the contract that the system must be in a quiescent state before running RTS, but it's also enforced
+by taking the session's checkpoint lock.
```
Comment (kommiharibabu) [inline src/docs/arch-rts.dox:221]: checkpoint lock is present in the connection. Change it to `connection`.

File: src/docs/arch-rts.dox [hunk 28]
```diff
@@ -110,4 +109,154 @@ Update list: U3 (30)\n
...
+
+Upon completion of RTS, a checkpoint is performed so that both in-memory and on-disk versions of
+data are the same. It's possible for users to opt out of this, for example if they intended to do a
+checkpoint shortly after RTS anyway.
+
+Eviction doesn't directly interact with RTS in any special way, but it's worth pointing out that RTS
+is likely to generate a large number of both clean and dirty pages, causing some amount of extra
+cache pressure. The clean pages can be evicted trivially, but the dirty pages will need to be
+written out, causing a potentially large amount of I/O and the potential for application threads to
+be forced into doing eviction work. Eviction triggers and targets apply the same way during RTS as
```
Comment (kommiharibabu) [inline src/docs/arch-rts.dox:231]: There will not be any application threads that are running in the WT when RTS is happening, so mentioning application threads doing eviction work may not be correct.

File: src/docs/arch-rts.dox (hunk 26 above)
Comment (korteland) [reply src/docs/arch-rts.dox:198]: It's the latter, and I think I wrote this the wrong way around for some reason - I believe it should say "... or if the page state is on-disk. If ..."

File: src/docs/arch-rts.dox [hunk 29]
```diff
@@ -110,4 +109,154 @@ Update list: U3 (30)\n
...
+to the stable timestamp.
+
+Timestamps of updates in the history store can have some non-obvious properties:
+
+* The start timestamp can be equal to the stop timestamp if the original update's commit timestamp
+  is in order.
+* During rollback of a prepared transaction, a history store update with a stop timestamp may not
+  get removed, leading to a duplicate record. RTS ignores these, and they're removed during the next
+  checkpoint.
+* If we've had to fix any timestamps as part of RTS, a newly inserted update with an older timestamp
```
[the remark this answers is not in the document]
Comment (korteland) [reply src/docs/arch-rts.dox:132]: This is referring to this comment in `rts_btree.c`:
```
* If we have fixed the missing timestamps, then the newer update reinserted with an older
* timestamp may have a durable timestamp that is smaller than the current stop durable
* timestamp.
```

File: src/docs/arch-rts.dox [hunk 30]
```diff
@@ -110,4 +109,154 @@ Update list: U3 (30)\n
...
+  may have a durable timestamp smaller than the current stop durable timestamp.
+* When a checkpoint writes a committed prepared update, it's possible for further updates on that
```
[the remark this answers is not in the document]
Comment (korteland) [reply src/docs/arch-rts.dox:134]: This is referring to a comment in `rts_btree.c` that would indicate it is possible:
```
* It is possible that there can be an update in the history store with a max stop timestamp
* in the middle of the same key updates. This occurs when the checkpoint writes the
* committed prepared update and further updates on that key including the history store
* changes before the transaction fixes the history store update to have a proper stop
* timestamp. It is a rare scenario.
```

It's possible I misunderstood the comment?

File: src/docs/arch-rts.dox [hunk 31]
```diff
@@ -110,4 +109,154 @@ Update list: U3 (30)\n
...
+Once RTS has finished, we roll back the global durable timestamp to the stable timestamp, since if
+there are no unstable updates, the global durable timestamp can, by definition, be moved backwards
+to the stable timestamp.
+
+Timestamps of updates in the history store can have some non-obvious properties:
+
+* The start timestamp can be equal to the stop timestamp if the original update's commit timestamp
+  is in order.
+* During rollback of a prepared transaction, a history store update with a stop timestamp may not
+  get removed, leading to a duplicate record. RTS ignores these, and they're removed during the next
```
[the remark this answers is not in the document]
Comment (korteland) [reply src/docs/arch-rts.dox:130]: This is referring to a comment in `rts_btree.c`:
```

* It is possible during a prepared transaction rollback, the history store update that have
* its own stop timestamp doesn't get removed leads to duplicate records in history store
* after further operations on that same key. Rollback to stable should ignore such records
* for timestamp ordering verification.
```

After re-reading it, I think I might've misunderstood - is the comment saying that we ignore them only for the purposes of verifying timestamp ordering? If so, I'll remove this point.

File: src/docs/arch-rts.dox (hunk 28 above)
Comment (korteland) [reply src/docs/arch-rts.dox]: Thanks - fixed in 5d27bf2.

File: src/docs/arch-rts.dox (hunk 27 above)
Comment (korteland) [reply src/docs/arch-rts.dox]: Thanks - fixed in 5d27bf2.

File: src/docs/arch-rts.dox (hunk 26 above)
Comment (korteland) [reply src/docs/arch-rts.dox]: Draft fix in 5d27bf2, let me know if it's OK.

File: src/docs/arch-rts.dox [hunk 32]
```diff
@@ -110,4 +109,154 @@ Update list: U3 (30)\n
...
+read an update to move it to the data store.
+
+The deletion code is called from the general update aborting code, where we might need to delete the
+first stable update we found when it came from the history store. There's also a somewhat subtle
+case where a new key can be added to a page, and then the page checkpointed, leading to a situation
+where we have updates to the key in the history store but not the disk image.
+
+History store deletions are done using a standard history store cursor, starting at the highest
+possible start timestamp, and removing keys from newest to oldest until we hit an update with a
+start timestamp larger than the rollback timestamp.
```
[the remark this answers is not in the document]
Comment (korteland) [reply src/docs/arch-rts.dox]: Thanks, fixed in 5d27bf2.

File: src/docs/arch-rts.dox [hunk 33]
```diff
@@ -110,4 +109,154 @@ Update list: U3 (30)\n
...
+records, we check that the start time of a newly encountered record is greater than or equal to the
+previous record. We also validate that the timestamps in the key and the cell are the same (where
+possible).
+
+Fast-truncate is worth mentioning briefly. If the page has a page delete structure, we can look at
+its durable timestamp, much the same as we would a normal page.
+
+@section interaction-with-transaction-ids Interaction with transaction IDs
+
+This is a little simpler than the timestamp rules - we use normal snapshot visibility in most of the
```
[the remark this answers is not in the document]
Comment (korteland) [reply src/docs/arch-rts.dox]: Thanks, fixed in 5d27bf2.

File: src/docs/arch-rts.dox [hunk 34]
```diff
@@ -110,4 +109,154 @@ Update list: U3 (30)\n
...
+there are no unstable updates, the global durable timestamp can, by definition, be moved backwards
+to the stable timestamp.
+
+Timestamps of updates in the history store can have some non-obvious properties:
+
+* The start timestamp can be equal to the stop timestamp if the original update's commit timestamp
+  is in order.
+* During rollback of a prepared transaction, a history store update with a stop timestamp may not
+  get removed, leading to a duplicate record. RTS ignores these, and they're removed during the next
+  checkpoint.
```
[the remark this answers is not in the document]
Comment (korteland) [reply src/docs/arch-rts.dox]: Oops, thanks for spotting this - fixed in 5d27bf2.

File: src/docs/arch-rts.dox (hunk 30 above)
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/docs/arch-rts.dox:134]: Thanks, now I understood.

File: src/docs/arch-rts.dox (hunk 29 above)
[the remark this answers is not in the document]
Comment (kommiharibabu) [reply src/docs/arch-rts.dox:132]: ok, the missing timestamps are referred to as out-of-order timestamps. This condition is no longer needed as we dropped supporting an out-of-order timestamp. We can either remove both this statement and the condition.

File: src/docs/arch-rts.dox (hunk 31 above)
[the remark this answers is not in the document]
Comment (korteland) [reply src/docs/arch-rts.dox]: Thanks for clarifying - I've fixed this bit, hopefully it's correct now.

**Review by kommiharibabu — APPROVED** [review]
LGTM

**Review by clarissecheah — APPROVED** [review]
LGTM!

File: src/docs/arch-rts.dox [hunk 35]
```diff
@@ -19,8 +19,7 @@ it meets one of the following conditions:
 applicable only during the restart phase).
 4. One of the checkpoint durable timestamps (start/stop) is greater than the stable timestamp.
 
-On the other hand, a table is skipped by RTS if it does not meet any of the previous criteria but
-does meet one of the following:
+RTS might choose to skip a table for any of these reasons (this list is not exhaustive):
 
 1. It is empty.
```
[the remark this answers is not in the document]
Comment (korteland) [reply src/docs/arch-rts.dox:24]: Good question - the comment is referring specifically to recovery (i.e. startup) and shutdown RTS. So those two cases don't apply when RTS called via the API (`conn->rollback_to_stable`).

[the remark this answers is not in the document]
Comment (clarissecheah) [reply src/docs/arch-rts.dox:24]: Ah yea, cheers. 👍
